Introduce 10.13 Libc

toonn opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Add missing headers.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

toonn committed:
apple-source-releases: Add 10.13.6 releases

The `applePackageMapping` function will allow us to keep expressions for
the newer releases separate from the originals where necessary, without
needing to change `generate-sdk-packages.sh`. If we ever want to support
multiple versions of source releases simultaneously we should consider
changing that to point to versioned subdirectories of all the packages.

Many of the source releases introduced here will need changes to the
expression to build successfully. These will be introduced one by one.

toonn committed:
darwin."10.13.6".Libc: Add missing headers to check

toonn committed:
darwin."10.13.6": Override the scope

Releases with dependencies on other releases, e.g., libresolv depends on
Libinfo, mDNSResponder and configdHeaders, were getting those releases
from pkgs.darwin rather than pkgs.darwin."10.13.6". By making the scope
of callPackage overridable we can manipulate which packages are used to
supply dependencies.

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/Libc/10.13.6.nix:
```diff
@@ -0,0 +1,27 @@
+    export DSTROOT=$out
+    export PUBLIC_HEADERS_FOLDER_PATH=include
+    export PRIVATE_HEADERS_FOLDER_PATH=include
+    bash xcodescripts/headers.sh
```
```suggestion
    FEATURE_LEGACY_RUNE_APIS=1 FEATURE_LEGACY_UTMP_APIS=1 bash xcodescripts/headers.sh
```

reckenrode replied:
That will ensure `rune.h` and `utmp.h` are installed. These are included in the Apple SDKs but not installed with the libc sources by default. The lack of `utmp.h` broke inettools during the LLVM 16 update. It has a workaround, but I wanted to fix libc to install the missing header.

philiptaron commented:
How can I help get this from draft to merged?

toonn committed:
darwin."10.13.6".Libc: Adopt

toonn commented:
@philiptaron, I'm marking the first batch of PRs as ready for review.

philiptaron commented:
I'll take a look today and get my Macs heating up.

philiptaron reviewed (commented):
I get the same error trying to build this as ofBorg: https://gist.github.com/GrahamcOfBorg/0bf469b9c307769b368a98965aad781d

toonn commented on pkgs/os-specific/darwin/apple-source-releases/default.nix:
```diff
@@ -221,14 +221,20 @@ let
   developerToolsPackages_11_3_1 = import ./developer-tools-11.3.1.nix { inherit applePackage'; };
 
-  applePackage' = namePath: version: sdkName: sha256:
+  applePackage'' = extra: namePath: version: sdkName: sha256:
```
We rely on partial application for the first two arguments so the best we could do is have an attrset for version, sdkName and sha256. Not sure that's worth it for a helper that is only used in the source release expressions.

philiptaron replied:
Yeah, just a suggestion. All that partial application is changeable. Just noting that at five arguments, this is pretty unwieldy.

philiptaron commented:
I'm working on getting the eval failure repro for you, @toonn, just running into "no SSH server on my Mac" issues.

toonn commented on pkgs/os-specific/darwin/apple-source-releases/default.nix:
```diff
@@ -321,4 +327,15 @@ developerToolsPackages_11_3_1 // macosPackages_11_0_1 // {
+                      }."${namePath}"
+                      or namePath;
+                 in import ./macos-10.13.6.nix
+                      { applePackage' = applePackageMapping; };
```
Force pushed with nixfmt formatting.

philiptaron commented:
Repro: `nix-build --expr 'let pkgs = import ./. { }; in builtins.attrValues pkgs.darwin."10.13.6"'`

toonn commented:
`nix-build --pure --arg system '"x86_64-darwin"' -A darwin.'"10.13.6"'.Libc` works though, no? That attrset isn't actually a derivation so it's not meant to be built. Not sure why it's missing the xpc argument when called that way but I'm not sure it matters.

toonn committed:
darwin."10.13.6": Only include working source releases

Some of the source releases build without changes to their expressions,
others require changes or the passing in of exceptional arguments. To
avoid evaluation of the entire "10.13.6" subset causing evaluation
errors we can include just the functional source releases.

toonn commented:
@philiptaron, with the new commits it should be fine to evaluate the entire "10.13.6" attrset now.

philiptaron reviewed (approved):
It does indeed all build for me. Thank you!

smancill commented:
Is this still ready to be merged or something has changed meanwhile?

reckenrode commented:
I’m not sure how this intersects with the Darwin refactor that’s coming. Some of the source releases are being updated and carried forward, but many others are being dropped.

toonn commented:
It mostly depends on what we want to do. Unless the source releases are being moved to by-name in the refactor there's no harm in having the 10.13 source builds.
However, since we'll be moving stdenv to the SDK, be that 10.12, 10.13 or more recent, there's not much use to them as-is I think. The interesting work going forward would be seeing what parts of the SDK we can build and overlay, creating a partially from-source frankenSDK.

reckenrode commented:
> It mostly depends on what we want to do. Unless the source releases are being moved to by-name in the refactor there's no harm in having the 10.13 source builds.

In my current branch, the source releases I’m using have been moved to the `by-name` hierarchy. The other ones are being left in place as stubs and will eventually be removed. Private headers can be copied from the source release source. There is a helper on the SDK package to get the source associated with that SDK (`apple-sdk.sourceRelease <repo name>`). That doesn’t always work (because the source releases are a mess), but it’s often good enough.

There are a few exceptions. darwin.bsdmake has been switched to bmake, which is the non-Apple distribution of BSD Make. darwin.ICU is (probably) going to provide a stub `libicucore.dylib` that reeexports icu’s dylibs. I haven’t gotten to that part yet. The rest of the source releases are being updated to the latest versions (with patches as necessary).

> However, since we'll be moving stdenv to the SDK, be that 10.12, 10.13 or more recent, there's not much use to them as-is I think. The interesting work going forward would be seeing what parts of the SDK we can build and overlay, creating a partially from-source frankenSDK.

When I first started working on it, I tried to pull headers from the source releases. I got to xnu and stopped because of differences. Having to patch the source releases with headers from the SDK seemed like it defeated the point. If someone does want to create an apple-sdk_src, the infrastructure is still in place to do that. It’s currently only used for AvailabilityVersions, but it should work with any designated package that has a passthru `injectFramework` function.
